horovod/spark/common/util.py

```diff
@@ -0,0 +1,518 @@
+    if validation_split and validation_col:
+        raise ValueError("can only specify one of validation_col and validation_split")
+
+    num_partitions = num_processes * partitions_per_process
+    if verbose:
+        print('num_partitions={}'.format(num_partitions))
+
+    train_data_path = store.get_train_data_path()
+    val_data_path = store.get_val_data_path()
+    if verbose:
+        print('train_data_path={}'.format(train_data_path))
+        print('val_data_path={}'.format(val_data_path))
+
+    for col in label_columns:
+        if col not in df.columns:
+            raise ValueError('Label column {} does not exist in this DataFrame'.format(col))
+
+    if feature_columns is None:
+        feature_columns = [col for col in df.columns if col not in set(label_columns)]
+
+    dataframe_hash = df.__hash__()
+
+    global _training_cache
+    key = (dataframe_hash, validation_split, validation_col, train_data_path, val_data_path)
+    with _training_cache.lock:
+        if _training_cache.is_cached(key):
+            entry = _training_cache.get(key)
+            train_rows = entry.train_rows
+            val_rows = entry.val_rows
+            metadata = entry.metadata
+            avg_row_size = entry.avg_row_size
+
+            if verbose:
+                print('using cached dataframes for key: {}'.format(key))
+                print('train_rows={}'.format(train_rows))
+                print('val_rows={}'.format(val_rows))
+        else:
+            if verbose:
+                print('writing dataframes')
+
+            schema_cols = feature_columns + label_columns
+            if sample_weight_col:
+                schema_cols.append(sample_weight_col)
+            if validation_col:
+                schema_cols.append(validation_col)
+            df = df[schema_cols]
+
+            metadata = _get_metadata(df)
+
+            to_petastorm = to_petastorm_fn(schema_cols, metadata)
+            train_df = df.rdd.map(to_petastorm).toDF()
+
+            val_df = None
+            validation_ratio = validation_split
+            if validation_split > 0:
+                train_df, val_df = train_df.randomSplit([1.0 - validation_split, validation_split])
+            elif validation_col:
+                val_df = train_df.filter(f.col(validation_col) > 0).drop(validation_col)
+                train_df = train_df.filter(f.col(validation_col) == 0).drop(validation_col)
+
+                # Approximate ratio of validation data to training data for proportionate scale
+                # of partitions
+                timeout_ms = 1000
+                confidence = 0.90
+                train_rows = train_df.rdd.countApprox(timeout=timeout_ms, confidence=confidence)
+                val_rows = val_df.rdd.countApprox(timeout=timeout_ms, confidence=confidence)
+                validation_ratio = val_rows / (val_rows + train_rows)
+
+            train_partitions = max(int(num_partitions * (1.0 - validation_ratio)),
```

Comment:
I think you should have at least `num_processes` partitions, not `partitions_per_process`, otherwise you are not utilizing all your executors / processes. Same for `val_partitions`.

Reply:
Yes, good catch.